src/sources/kafka.rs

```diff
@@ -352,14 +361,36 @@ fn create_consumer(
     }
 
     let consumer = client_config
-        .create_with_context::<_, StreamConsumer<_>>(KafkaStatisticsContext)
+        .create_with_context::<_, StreamConsumer<_>>(CustomContext::default())
         .context(KafkaCreateSnafu)?;
     let topics: Vec<&str> = config.topics.iter().map(|s| s.as_str()).collect();
     consumer.subscribe(&topics).context(KafkaSubscribeSnafu)?;
 
     Ok(consumer)
 }
 
+#[derive(Default)]
+struct CustomContext {
+    stats: kafka::KafkaStatisticsContext,
+    finalizer: OnceCell<Arc<OrderedFinalizer<FinalizerEntry>>>,
+}
+
+impl ClientContext for CustomContext {
+    fn stats(&self, statistics: Statistics) {
+        self.stats.stats(statistics)
+    }
+}
+
+impl ConsumerContext for CustomContext {
+    fn post_rebalance(&self, rebalance: &Rebalance) {
+        if matches!(rebalance, Rebalance::Revoke(_)) {
+            if let Some(finalizer) = self.finalizer.get() {
+                finalizer.flush();
+            }
+        }
```

Comment:
The one thing I don't remember from when I was poking at the unit test you wrote for this behavior is whether or not all consumers have their partitions revoked and then re-assigned during a rebalance event, or if only the revocations necessary to balance the partitions will occur.

If it's the latter, then we might actually be breaking finalization for messages in partition that we haven't had revoked from us.

Reply:
It is the latter, based on my experiments in testing. You are correct, we do not minimize which finalizations we drop from the queue. However, given streaming data from Kafka, the next acknowledged new piece of data will correctly re-set the offset. I'm a little nervous to make this effort any more complex given our difficulty testing it ourselves.

Reply:
👍 I think the end result is some event duplication which is better than loss. We should make an issue just to track following up on this though.

Reply:
Totally understand about wanting to limit complexity.

I think I would quibble over the "next acknowledged new piece of data" bit, though.  Here's how I'm thinking about this:

- a consumer is processing two partition, and has two outstanding/in-flight messages, one from each partition
- a rebalance happens, and the consumer loses one partition, but resets all in-flight finalizers
- this is fine for the partition that it actually lost
- the consumer pulls the next message from the partition it still has, and the finalizer runs and updates the offset for that partition
- this happens _before_ the previous message for that partition is finalized: either due to latency or due to the actual message failing to process

Am I wrong or are we not effectively telling Kafka that all the messages we had yet to acknowledge on the partitions we still control are now acknowledged?

Reply:
I agree with your timeline, but there is one missing step: Vector is killed before that previous message is finalized. *Then* that unfinalized message is lost.

I agree this is worth correcting. However, given that the approach in this PR is an improvement to the current state, has been in review state for over a month, and the corrections to resolve this could be non-trivial, I would prefer to complete this PR as-is and add another issue to fix the above correctness hole.

Reply:
Oh, absolutely.  I mostly just wanted to make sure we were on the same page, or correct my thinking if it was in error, since we're already reviewing this code and might as well get it as solid as we reasonably can. :)

+1 to a follow-up issue.